chat/actions.go

```
@@ -0,0 +1,272 @@
+			if _, ok := o["link"]; ok {
+				// Is a public image, just append it
+				objects = append(objects, o)
+			} else {
+				fo := NewObject(s.FileInteractor)
+				err := fo.BuildFromData(o["data"].([]byte), o["name"].(string), o["mime"].(string))
+				if err == nil {
+					objects = append(objects, fo.ToPayload())
+				} else {
+					log.Println(err.Error())
+				}
+			}
+		}
+		payload["objects"] = objects
+	}
+
+	s.send(recipients, payload)
+
+	return NewMessage(s, recipients, payload)
+}
+
+// Delivered sends a message to confirm a list of messages (identified by it's cids)
+// have been delivered.
+func (s *Service) Delivered(recipients []string, cids []string, gid string) {
+	s.confirm("delivered", recipients, cids, gid)
+}
+
+// Read sends a message to confirm a list of messages (identified by it's cids)
+// have been read.
+func (s *Service) Read(recipients []string, cids []string, gid string) {
+	s.confirm("read", recipients, cids, gid)
+}
+
+// Edit changes the body of a previous message.
+func (s *Service) Edit(recipients []string, cid string, body string, gid string) {
+	p := map[string]interface{}{
+		"typ": "chat.message.edit",
+		"cid": cid,
+		"msg": body,
+	}
+
+	if len(gid) > 0 {
+		p["gid"] = gid
+	}
+
+	s.send(recipients, p)
+}
+
+// Delete deletes previous messages.
+func (s *Service) Delete(recipients []string, cids []string, gid string) {
+	p := map[string]interface{}{
+		"typ":  "chat.message.delete",
+		"cids": cids,
+	}
+
+	if gid != "" {
+		p["gid"] = gid
+	}
+
+	s.send(recipients, p)
+}
+
+// Invite sends a group invitation to a list of members.
+func (s *Service) Invite(gid string, name string, members []string, opts map[string]interface{}) {
+	p := map[string]interface{}{
+		"typ":     "chat.invite",
+		"gid":     gid,
+		"name":    name,
+		"members": members,
+		"aud":     gid,
+	}
+
+	if _, ok := opts["data"]; ok {
+		fo := NewObject(s.FileInteractor)
+		err := fo.BuildFromData(opts["data"].([]byte), "", opts["mime"].(string))
```

This should really return an error to the user rather than silently failing. 

Also, how come you're passing `FileInteractor` to each object, instead of:
```
// getting an objects data
data, err := s.GetObject(obj) // invokes FileInteractor

// creating an object
obj, err := s.CreateObject("name", io.NewReader([]byte("data")))
```

That way, you don't have to expose `FileInteractor`, which is an implementation detail